sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -312,45 +270,460 @@ class reducer<T, BinaryOperation,
               (Space == access::address_space::global_space ||
                Space == access::address_space::local_space)>
   atomic_combine(_T *ReduVarPtr) const {
-    atomic_ref<T, memory_order::relaxed, getMemoryScope<Space>(), Space>(
-        *multi_ptr<T, Space>(ReduVarPtr))
-        .fetch_and(MValue);
+    auto reducer = static_cast<const Reducer *>(this);
+    for (size_t E = 0; E < Extent; ++E) {
+      atomic_ref<T, memory_order::relaxed, getMemoryScope<Space>(), Space>(
+          multi_ptr<T, Space>(ReduVarPtr)[E])
+          .fetch_and(reducer->getElement(E));
+    }
+  }
+
+  /// Atomic MIN operation: *ReduVarPtr = sycl::minimum(*ReduVarPtr, MValue);
+  template <access::address_space Space = access::address_space::global_space,
+            typename _T = T, class _BinaryOperation = BinaryOperation>
+  enable_if_t<std::is_same<typename remove_AS<_T>::type, T>::value &&
+              IsReduOptForFastAtomicFetch<T, _BinaryOperation>::value &&
+              sycl::detail::IsMinimum<T, _BinaryOperation>::value &&
+              (Space == access::address_space::global_space ||
+               Space == access::address_space::local_space)>
+  atomic_combine(_T *ReduVarPtr) const {
+    auto reducer = static_cast<const Reducer *>(this);
+    for (size_t E = 0; E < Extent; ++E) {
+      atomic_ref<T, memory_order::relaxed, getMemoryScope<Space>(), Space>(
+          multi_ptr<T, Space>(ReduVarPtr)[E])
+          .fetch_min(reducer->getElement(E));
+    }
+  }
+
+  /// Atomic MAX operation: *ReduVarPtr = sycl::maximum(*ReduVarPtr, MValue);
+  template <access::address_space Space = access::address_space::global_space,
+            typename _T = T, class _BinaryOperation = BinaryOperation>
+  enable_if_t<std::is_same<typename remove_AS<_T>::type, T>::value &&
+              IsReduOptForFastAtomicFetch<T, _BinaryOperation>::value &&
+              sycl::detail::IsMaximum<T, _BinaryOperation>::value &&
+              (Space == access::address_space::global_space ||
+               Space == access::address_space::local_space)>
+  atomic_combine(_T *ReduVarPtr) const {
+    auto reducer = static_cast<const Reducer *>(this);
+    for (size_t E = 0; E < Extent; ++E) {
+      atomic_ref<T, memory_order::relaxed, getMemoryScope<Space>(), Space>(
+          multi_ptr<T, Space>(ReduVarPtr)[E])
+          .fetch_max(reducer->getElement(E));
+    }
+  }
+};
+
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
```

Comment:
Would you please copy-paste-adjust the comment section from L342 (from the the next specialization) to here?

Reply:
Added a comment in 657ff53321c5.